test/CMakeLists.txt

```diff
@@ -64,6 +64,45 @@ if(MIOPEN_BACKEND_HIP AND NOT MIOPEN_TEST_HIP_NOGPU)
     set(MIOPEN_TEST_HIP TRUE)
 endif()
 
+# Detect GPU type for testing.
+# For HIP_NOGPU backend, GPU detection is not required and should be disabled.
+# Also we do not detect GPU when target GPU for testing is specified explicitly.
+set(MIOPEN_TEST_GPU_DETECTION_FAILED FALSE)
+if(NOT (MIOPEN_TEST_VEGA OR MIOPEN_TEST_GFX908 OR MIOPEN_TEST_HIP_NOGPU))
+    find_program(ROCMINFO
+        NAMES rocminfo
+        PATHS
+            /opt/rocm
+            ${CMAKE_INSTALL_PREFIX}
+        PATH_SUFFIXES
+            /bin
+    )
+    message(STATUS "rocminfo utility: ${ROCMINFO}")
+    if(ROCMINFO)
+        execute_process (
+            COMMAND ${ROCMINFO}
+            OUTPUT_VARIABLE ROCMINFO_OUTPUT
+            RESULT_VARIABLE ROCMINFO_EXIT_STATUS
+        )
+        if(NOT ROCMINFO_EXIT_STATUS EQUAL 0)
+            message(WARNING "ROCMINFO FAILED, GPU TYPE UNKNOWN. Manually set respective MIOPEN_TEST_GFX* CMake variable to specify target GPU for testing.")
+            set(MIOPEN_TEST_GPU_DETECTION_FAILED TRUE)
```

Comment:
Maybe this could be done using a function rather than setting a cmake variable?

Reply:
Yes, this can be done, but the function can be called only after `add_custom_target(check COMMAND ${CMAKE_CTEST_COMMAND} --output-on-failure -C ${CMAKE_CFG_INTDIR})`. The `MIOPEN_TEST_GPU_DETECTION_FAILED` variable which holds the status of detection provides more flexibility. Let me keep it as is for now.

Reply:
Sure